File: src/sokoenginepy/common/trioban_tessellation.py
from __future__ import annotations

from typing import Final, Mapping, Tuple

from .tile_shape import TileShape
from .characters import Characters
from .config import Config
from .direction import Direction
from .graph_type import GraphType
from .tessellation import index_1d, index_column, index_row, is_on_board_2d
from .tessellation_impl import PusherStepData, TessellationImpl


class TriobanTessellation(TessellationImpl):
    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    _CHR_TO_PUSHER_STEP: Final[Mapping[str, PusherStepData]] = {
        Characters.l: PusherStepData(Direction.LEFT, False),
        Characters.L: PusherStepData(Direction.LEFT, True),
        Characters.r: PusherStepData(Direction.RIGHT, False),
        Characters.R: PusherStepData(Direction.RIGHT, True),
        Characters.n: PusherStepData(Direction.NORTH_EAST, False),
        Characters.N: PusherStepData(Direction.NORTH_EAST, True),
        Characters.u: PusherStepData(Direction.NORTH_WEST, False),
        Characters.U: PusherStepData(Direction.NORTH_WEST, True),
        Characters.d: PusherStepData(Direction.SOUTH_EAST, False),
        Characters.D: PusherStepData(Direction.SOUTH_EAST, True),
        Characters.s: PusherStepData(Direction.SOUTH_WEST, False),
        Characters.S: PusherStepData(Direction.SOUTH_WEST, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Mapping[PusherStepData, str]] = {
        v: k for k, v in _CHR_TO_PUSHER_STEP.items()
    }

    @property
    def graph_type(self) -> GraphType:
        return GraphType.DIRECTED_MULTI

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        triangle_points_down = (
            self.tile_shape(position, board_width, board_height)
            == TileShape.TRIANGLE_DOWN
        )
        row = index_row(position, board_width)
        column = index_column(position, board_width)

        dx, dy = 0, 0
        if direction == Direction.LEFT:
            dy = 0
            dx = -1
        elif direction == Direction.RIGHT:
            dy = 0
            dx = 1
        elif direction == Direction.NORTH_EAST:
            if triangle_points_down:
                dy = -1
                dx = 0
            else:
                dy = 0
                dx = 1
        elif direction == Direction.NORTH_WEST:
            if triangle_points_down:
                dy = -1
                dx = 0
            else:
                dy = 0
                dx = -1
        elif direction == Direction.SOUTH_EAST:
            if triangle_points_down:
                dy = 0
                dx = 1
            else:
                dy = 1
                dx = 0
        elif direction == Direction.SOUTH_WEST:
            if triangle_points_down:
                dy = 0
                dx = -1
            else:
                dy = 1
                dx = 0
        else:
            raise ValueError(
                f"Unsupported direction {direction} for {self.__class__.__name__}"
            )

        row += dy
        column += dx

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS

    def tile_shape(
        self, position: int, board_width: int, board_height: int
    ) -> TileShape:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)
        return (
            TileShape.TRIANGLE_DOWN
            if (column + (row % 2)) % 2 == 0
            else TileShape.DEFAULT
        )

File: src/libsokoengine/common/trioban_tessellation.cpp
/// @file
#include "trioban_tessellation.hpp"

#include "characters.hpp"
#include "pusher_step.hpp"

using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace implementation {

static const directions_t TRI_LEGAL_DIRECTIONS = {
  Direction::LEFT,
  Direction::RIGHT,
  Direction::NORTH_EAST,
  Direction::NORTH_WEST,
  Direction::SOUTH_EAST,
  Direction::SOUTH_WEST};

const directions_t &TriobanTessellation::legal_directions() const {
  return TRI_LEGAL_DIRECTIONS;
}

position_t TriobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  if (!is_on_board_1d(position, width, height))
    return Config::NO_POS;
  position_t row = index_y(position, width), column = index_x(position, width);
  int8_t     dy, dx;
  bool       tpd = tile_shape(position, width, height) == TileShape::TRIANGLE_DOWN;

  switch (direction) {
    case Direction::LEFT:
      dy = 0;
      dx = -1;
      break;
    case Direction::RIGHT:
      dy = 0;
      dx = 1;
      break;
    case Direction::NORTH_EAST:
      if (tpd) {
        dy = -1;
        dx = 0;
      } else {
        dy = 0;
        dx = 1;
      }
      break;
    case Direction::NORTH_WEST:
      if (tpd) {
        dy = -1;
        dx = 0;
      } else {
        dy = 0;
        dx = -1;
      }
      break;
    case Direction::SOUTH_EAST:
      if (tpd) {
        dy = 0;
        dx = 1;
      } else {
        dy = 1;
        dx = 0;
      }
      break;
    case Direction::SOUTH_WEST:
      if (tpd) {
        dy = 0;
        dx = -1;
      } else {
        dy = 1;
        dx = 0;
      }
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + to_str(direction) + " for TriobanTessellation!"
      );
  }

  row += dy;
  column += dx;
  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

PusherStep TriobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case Characters::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case Characters::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);

    case Characters::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case Characters::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);

    case Characters::n:
      return PusherStep(Direction::NORTH_EAST, Config::NO_ID);
    case Characters::N:
      return PusherStep(Direction::NORTH_EAST, Config::DEFAULT_ID);

    case Characters::u:
      return PusherStep(Direction::NORTH_WEST, Config::NO_ID);
    case Characters::U:
      return PusherStep(Direction::NORTH_WEST, Config::DEFAULT_ID);

    case Characters::d:
      return PusherStep(Direction::SOUTH_EAST, Config::NO_ID);
    case Characters::D:
      return PusherStep(Direction::SOUTH_EAST, Config::DEFAULT_ID);

    case Characters::s:
      return PusherStep(Direction::SOUTH_WEST, Config::NO_ID);
    case Characters::S:
      return PusherStep(Direction::SOUTH_WEST, Config::DEFAULT_ID);

    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in TriobanTessellation!"
      );
  }
}

char TriobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? Characters::L : Characters::l;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? Characters::R : Characters::r;
    case Direction::NORTH_EAST:
      return rv.is_push_or_pull() ? Characters::N : Characters::n;
    case Direction::NORTH_WEST:
      return rv.is_push_or_pull() ? Characters::U : Characters::u;
    case Direction::SOUTH_EAST:
      return rv.is_push_or_pull() ? Characters::D : Characters::d;
    case Direction::SOUTH_WEST:
      return rv.is_push_or_pull() ? Characters::S : Characters::s;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + to_str(rv.direction())
        + " in TriobanTessellation!"
      );
  }
}

TileShape TriobanTessellation::tile_shape(
  position_t pos, board_size_t width, board_size_t height
) const {
  position_t column = index_column(pos, width);
  position_t row    = index_row(pos, width);
  return (column + (row % 2)) % 2 == 0 ? TileShape::TRIANGLE_DOWN : TileShape::DEFAULT;
}

GraphType TriobanTessellation::graph_type() const { return GraphType::DIRECTED_MULTI; }

tile_shapes_t TriobanTessellation::tile_shapes() const {
  return tile_shapes_t{TileShape::DEFAULT, TileShape::TRIANGLE_DOWN};
}

} // namespace implementation
} // namespace sokoengine
